nexus/deployment/src/planner.rs

```diff
@@ -66,6 +80,41 @@ impl<'a> Planner<'a> {
                 continue;
             }
 
+            // Now we've established that the current blueprint _says_ there's
+            // an NTP zone on this system.  But we must wait for it to actually
+            // be there before we can proceed to add anything else.  Otherwise,
+            // we may wind up trying to provision this zone at the same time as
+            // other zones, and Sled Agent will reject requests to provision
+            // other zones before the clock is synchronized.
+            //
+            // Note that it's always possible that the NTP zone was added after
+            // this inventory was collected (in which case we'll erroneously
+            // choose to bail out here, but we'll pick it up again next time
+            // we're invoked).  It's conceivable that the NTP zone was removed
+            // after this inventory was collected (in which case we'd be making
+            // a wrong decision here).  However, we don't ever do this today.
+            // If we were to do something like that (maybe as part of upgrading
+            // the NTP zone or switching between an internal NTP vs. boundary
+            // NTP zone), we'll need to be careful how we do it to avoid a
+            // problem here.
+            let has_ntp_inventory = self
+                .inventory
+                .omicron_zones
+                .get(&sled_id)
+                .map(|sled_zones| {
+                    sled_zones.zones.zones.iter().any(|z| z.zone_type.is_ntp())
+                })
+                .unwrap_or(false);
+            if !has_ntp_inventory {
+                info!(
+                    &self.log,
+                    "parent blueprint contains NTP zone, but it's not in \
+                    inventory yet";
+                    "sled_id" => ?sled_id,
+                );
+                continue;
```

Comment:
A thought that occurred to me: is there some place in the blueprint, or elsewhere, that we should be storing these decisions and the reasons for them? Just having them in logs seems hard to backfill on in the futuree. I know eventually we want to build a full-fledged log, but I think an incremental step where we just store things like "here are all the places we skipped making a decision, and here's why" would be pretty useful.

Reply:
That's a good idea.  One thought is that we could have an enum of different events ("event" isn't quite the right word for it) -- basically one for each branch or decision made.  Then we could store a list of these somewhere.  So you might have `[ FoundNtpBlueprint, FoundNtpInventory, ProvisionedCrucible ]`.  But I'm not sure where would make sense for it.  We could stick an array in the Blueprint itself, or use a separate table, or assume that we can pull this out of the log...other ideas?  I think we should be sure that this is only for debugging.  I don't think we want programmatic consumers looking at this to try to pick apart what changes were part of a blueprint, despite it being tempting to do that.

Tangentially: I kind of expect that in the long run, we'll wind up breaking the planner into more orthogonal steps.  Originally I'd envisioned a sequence of trait objects that provide something like a `plan(previous: &Blueprint, next: &mut BlueprintBuilder)` method, and the (generic) planner would just run them in sequence and each one would make whatever modifications it wanted.  If we do wind up doing something like that, these methods could return a description of what choice(s) they made.  I'm not convinced we've got enough concrete impls here yet for this to be worthwhile though.